gensim/models/word2vec.py

```diff
@@ -474,22 +503,27 @@ def create_binary_tree(self):
 
             logger.info("built huffman tree with maximum node depth %i", max_depth)
 
-    def build_vocab(self, sentences, keep_raw_vocab=False):
+    def build_vocab(self, sentences, keep_raw_vocab=False, trim_rule=None):
         """
         Build vocabulary from a sequence of sentences (can be a once-only generator stream).
         Each sentence must be a list of unicode strings.
 
         """
-        self.scan_vocab(sentences)  # initial survey
-        self.scale_vocab(keep_raw_vocab)  # trim by min_count & precalculate downsampling
+        if trim_rule is not None:
```

Comment:
I'd be happy to do the Doc2Vec adaptation, but since Doc2Vec shares so much of the vocab-scanning code, and it looks like it's only a few lines, it'd be good to keep it at parity in the same PR. Looking at changes that'd require (mainly `Doc2Vec.scan_vocab()`), how about moving `_keep_vocab_item` into utils, to eliminate the need for a `_modified_trim_rule` wrapper and ensure parameters named `trim_rule` have the same signature/return-values wherever they are used?